interfaces/prompting/patterns.go

```diff
@@ -0,0 +1,335 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package prompting
+
+import (
+	"fmt"
+	"io"
+	"regexp"
+	"strings"
+
+	doublestar "github.com/bmatcuk/doublestar/v4"
+)
+
+// The default previously-expanded prefixes to which new patterns or expanded
+// groups are concatenated. This must be a slice containing the empty string,
+// since at the beginning of the pattern, we have only one prefix to which to
+// concatenate, and that prefix is the empty string. Importantly, this cannot
+// be an empty slice, since concatenating every entry in an empty slice with
+// every entry in a slice of expanded patterns would again result in an empty
+// slice.
+var defaultPrefixes = []string{""}
+
+// Limit the number of groups which are allowed to occur in a pattern.
+// This number includes all groups, including those nested or in series.
+const maxGroupsInPattern = 10
+
+// ExpandPathPattern expands all groups in the given path pattern.
+//
+// Groups are enclosed by '{' '}'. Returns a list of all the expanded path
+// patterns, or an error if the given pattern is invalid.
+func ExpandPathPattern(pattern string) ([]string, error) {
```

Comment:
How about we split the responsibility of parsing and rendering:

```go
package main

import (
	"bytes"
	"fmt"
)

// RenderConfig is a configuration of a render node.
type RenderConfig interface {
	// Next modifies the configuration to the next state returning true if more
	// configurations remain to be rendered.
	Next() bool
}

// RenderNode is a node which may be rendered in a particular configuration.
type RenderNode interface {
	// Render renders the given configuration to the buffer.
	Render(*bytes.Buffer, RenderConfig)
	// Config returns the initial render configuration.
	Config() RenderConfig
	// NumVariants returns the number of variants this node can be rendered as (recursively).
	NumVariants() int
}

func main() {
	tree := Seq{
		// From snap-confine apparmor profile:
		// /{,usr/}lib{,32,64,x32}/{,@{multiarch}/{,atomics/}}ld{-*,64}.so*,
		Literal("/"),
		Alt{
			Literal(""),
			Literal("usr/"),
		},
		Literal("lib"),
		Alt{
			Literal(""),
			Literal("32"),
			Literal("64"),
			Literal("x32"),
		},
		Literal("/"),
		Alt{
			Literal(""),
			Seq{
				Literal("@{multiarch}/"),
				Alt{
					Literal(""),
					Literal("atomics/"),
				},
			},
		},
		Literal("ld"),
		Alt{
			Literal("-*"),
			Literal("64"),
		},
		Literal(".so*"),
	}

	fmt.Printf("Parse tree: %#v\n", tree)

	variants := make([]string, 0, tree.NumVariants())
	fmt.Printf("Number of variants: %d\n", cap(variants))

	RenderAllVariants(tree, func(buf *bytes.Buffer) { variants = append(variants, buf.String()) })
	for i, v := range variants {
		fmt.Printf("%4d: %s\n", i, v)
	}
}

// RenderAllVariants renders subsequent variants to a reusable buffer.
func RenderAllVariants(n RenderNode, observe func(*bytes.Buffer)) {
	var buf bytes.Buffer

	c := n.Config()

	for {
		// TODO: change how buffer is handled, so that we only need to re-render
		// the part that was invalidated by the old config.
		buf.Truncate(0)
		n.Render(&buf, c)
		observe(&buf)
		if !c.Next() {
			break
		}
	}
}

// Seq is sequence of consecutive render nodes.
type Seq []RenderNode

func (n Seq) Render(buf *bytes.Buffer, conf RenderConfig) {
	for i := range n {
		n[i].Render(buf, conf.(*seqConfig).configAt[i])
	}
}

func (n Seq) NumVariants() int {
	num := 1

	for i := range n {
		num *= n[i].NumVariants()
	}

	return num
}

func (n Seq) Config() RenderConfig {
	cfg := &seqConfig{
		configAt: make([]RenderConfig, len(n)),
		seqNode:  n,
	}

	for i := range n {
		cfg.configAt[i] = n[i].Config()
	}

	return cfg
}

type seqConfig struct {
	configAt []RenderConfig
	seqNode  Seq
}

func (c *seqConfig) Next() bool {
	for i := len(c.configAt) - 1; i >= 0; i-- {
		if c.configAt[i].Next() {
			return true
		}

		c.configAt[i] = c.seqNode[i].Config()
	}

	return false
}

// Alt is a sequence of alternative render nodes.
type Alt []RenderNode

func (n Alt) Render(buf *bytes.Buffer, conf RenderConfig) {
	n[conf.(*altConfig).currentVariant].Render(buf, conf.(*altConfig).currentConfig)
}

func (n Alt) NumVariants() int {
	num := 0

	for i := range n {
		num += n[i].NumVariants()
	}

	return max(num, 1)
}

func (n Alt) Config() RenderConfig {
	if len(n) == 0 {
		return nil
	}

	return &altConfig{
		altNode:        n,
		currentVariant: 0,
		currentConfig:  n[0].Config(),
	}
}

type altConfig struct {
	altNode        Alt
	currentVariant int
	currentConfig  RenderConfig
}

func (c *altConfig) Next() bool {
	if c == nil {
		return false
	}

	if c.currentConfig.Next() {
		return true
	}

	c.currentVariant++

	if c.currentVariant >= len(c.altNode) {
		return false
	}

	c.currentConfig = c.altNode[c.currentVariant].Config()

	return true
}

// Literal is a render node with a literal string.
type Literal string

func (n Literal) Render(buf *bytes.Buffer, conf RenderConfig) {
	buf.WriteString(string(n))
}

func (n Literal) NumVariants() int {
	return 1
}

func (n Literal) Config() RenderConfig {
	return literalConfig{}
}

type literalConfig struct{}

func (literalConfig) Next() bool {
	return false
}
```

With this you can render the tree that is somewhat easy (at least easier) to parse.

Reply:
the suggestion is reasonable, but the naming is very confusing, especially "Config"

Reply:
I have used the word _config_ as rendering all cases explores the configuration space of the tree, visiting all the leaves and printing the entire path along the way.

I can rename that to something else, I think that is not the key problem.

@olivercalder shall we team up to try to adjust the recursive parser to build the tree and combine it with the tree renderer?